conkit/misc/selectalg.py

```diff
@@ -0,0 +1,172 @@
+"""Energy function templates for restraint generation"""
+
+__author__ = 'Felix Simkovic'
+__date__ = '13 Aug 2018'
+__version__ = '1.0'
+
+import inspect
+import numpy as np
+
+
+class SubselectionAlgorithm(object):
+    """A class to collect all subselection algorithms"""
+
+    @staticmethod
+    def _numpify(data):
+        """Convert a Python array to a Numpy array"""
+        if type(data).__module__ == np.__name__:
+            return data
+        else:
+            return np.asarray(data)
+
+    @staticmethod
+    def cutoff(data, cutoff=0.287):
+        """A cutoff-defined subselection algorithm
+
+        Description
+        -----------
+        This algorithm removes a decoy, if its score is l
+        ess than the cutoff.
+
+        Parameters
+        ----------
+        data : list, tuple
+           A 1D array of scores
+        cutoff : float, optional
+           The cutoff of keeping decoys
+
+        Returns
+        -------
+        list
+           The decoy indices to keep
+        list
+           The decoy indices to throw
+
+        """
+        data = SubselectionAlgorithm._numpify(data)
+        keep = np.where(data >= cutoff)[0]
+        throw = np.where(data < cutoff)[0]
+        return keep.tolist(), throw.tolist()
+
+    @staticmethod
+    def linear(data, cutoff=0.5):
+        """A linearly-defined subselection algorithm
+
+        Description
+        -----------
+        This algorithm removes the worst 500 decoys.
+
+        Parameters
+        ----------
+        data : list, tuple
+           A 1D array of scores
+        cutoff : float, optional
+           The porportion of the total number of decoys to keep
+
+        Returns
+        -------
+        list
+           The decoy indices to keep
+        list
+           The decoy indices to throw
+
+        """
+        sorted_indices = SubselectionAlgorithm._numpify(data).argsort()[::-1]
+        point = np.ceil(sorted_indices.shape[0] * cutoff).astype(np.int)
+        keep = sorted_indices[:point]
+        throw = sorted_indices[point:]
+        return keep.tolist(), throw.tolist()
+
+    @staticmethod
+    def scaled(data, cutoff=0.5):
+        """A scaling-defined subselection algorithm
+
+        Description
+        -----------
+        This algorithm removes a decoy, if its scaled score
+        is less than 0.5. The scaled score is calculated by
+        dividing the precision score by the average of the
+        set.
+
+        Parameters
+        ----------
+        data : list, tuple
+           A 1D array of scores
+        cutoff : float, optional
+           The cutoff of keeping decoys
+
+        Returns
+        -------
+        list
+           The decoy indices to keep
+        list
+           The decoy indices to throw
+
+        """
+        data = SubselectionAlgorithm._numpify(data)
+        data_scaled = data / np.mean(data)
+        keep = np.where(data_scaled >= cutoff)[0]
+        throw = np.where(data_scaled < cutoff)[0]
```

Comment:
Could replace this with a call to the cutoff function above which seems to do the same thing.

Reply:
Good spot, thanks